DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,222 @@
+﻿namespace DigitalLearningSolutions.Data.Services
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using ClosedXML.Excel;
+    using DigitalLearningSolutions.Data.DataServices;
+    using DigitalLearningSolutions.Data.DataServices.UserDataService;
+    using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Models.DelegateUpload;
+    using DigitalLearningSolutions.Data.Models.Register;
+    using DigitalLearningSolutions.Data.Models.User;
+    using Microsoft.AspNetCore.Http;
+
+    public interface IDelegateUploadFileService
+    {
+        public BulkUploadResult ProcessDelegatesFile(IFormFile file, int centreId, DateTime? welcomeEmailDate = null);
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate = null);
+        public IXLTable OpenDelegatesTable(IFormFile file);
+    }
+
+    public class DelegateUploadFileService : IDelegateUploadFileService
+    {
+        private readonly IJobGroupsDataService jobGroupsDataService;
+        private readonly IRegistrationDataService registrationDataService;
+        private readonly IUserDataService userDataService;
+
+        public DelegateUploadFileService(
+            IJobGroupsDataService jobGroupsDataService,
+            IUserDataService userDataService,
+            IRegistrationDataService registrationDataService
+        )
+        {
+            this.userDataService = userDataService;
+            this.registrationDataService = registrationDataService;
+            this.jobGroupsDataService = jobGroupsDataService;
+        }
+
+        public BulkUploadResult ProcessDelegatesFile(IFormFile file, int centreId, DateTime? welcomeEmailDate)
+        {
+            var table = OpenDelegatesTable(file);
+            return ProcessDelegatesTable(table, centreId, welcomeEmailDate);
+        }
+
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate)
+        {
+            var jobGroupIds = jobGroupsDataService.GetJobGroupsAlphabetical().Select(item => item.id).ToList();
+            var (registered, updated, skipped) = (0, 0, 0);
+            var errors = new List<(int, BulkUploadResult.ErrorReason)>();

```

Comment:
I wonder whether you could store these on the delegateRow as nullable error and Status (enum - not processed, skipped, registered, updated)? Then, at the end of processing you could count the rows with different statuses and map the delegateRows to errors. Currently, passing the registered, updated, skipped counts around is a bit difficult to follow and could end up with a bug where for example single row increases two counts at once. 

You could also do it in a more functional way and map the DelegateTableRow to ProcessedDelegateTableRow or similar (returned by ProcessRow function), which has information about the row number, status and nullable error. Then use the mapped objects to get the final counts, errors, etc. :)